qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -0,0 +1,188 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Optimize annotated operations on a circuit."""
+
+from typing import Optional, List, Tuple
+
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.circuit.annotated_operation import AnnotatedOperation, _canonicalize_modifiers
+from qiskit.circuit import EquivalenceLibrary, ControlledGate, Operation, ControlFlowOp
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes.utils import control_flow
+from qiskit.transpiler.target import Target
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class OptimizeAnnotated(TransformationPass):
+    """Optimization pass on circuits with annotated operations."""
+
+    def __init__(
+        self,
+        target: Optional[Target] = None,
+        equivalence_library: Optional[EquivalenceLibrary] = None,
+        basis_gates: Optional[List[str]] = None,
+    ):
+        """
+        OptimizeAnnotated initializer.
+
+        Args:
+            target: Optional, the backend target to use for this pass.
+            equivalence_library: The equivalence library used
+                (instructions in this library will not be optimized by this pass).
+            basis_gates: Optional, target basis names to unroll to, e.g. `['u3', 'cx']`
+                (instructions in this list will not be optimized by this pass).
+                Ignored if ``target`` is also specified.
```

Comment:
Do you think having an explicit option to recurse into definitions is useful or not? Right now it's only done implicitly based on the presence of a `Target` or basis gates. But do you think there is value in having an override for it?

Reply:
After thinking about this, such an option could definitely be useful and is added in [fb7e279](https://github.com/Qiskit/qiskit/pull/11476/commits/fb7e27901e4350e358fb8877c4352d083598bf0a).

The current logic is the following (please see if you agree with this):

* If neither ``basis_gates`` nor ``target`` is specified, then we do not recursively descent into gate definitions, regardless of the ``recurse`` option. This is consistent with the behavior of ``HighLevelSynthesis`` pass (and previously of ``UnrollCustomDefinitions`` pass) which in this case does not descend into the definitions either, i.e. we shouldn't optimize what we are not going to synthesize. 
* If either ``basis_gates`` or ``target`` is specified, and ``recurse`` has the default value of ``True``, then we do the recursion.
* But we can override the recursion by setting ``recurse`` to ``False``. 

This also ties in to the question of where this pass should sit in the transpilation pipeline. I was also thinking about the `init` stage (just before running high level synthesis). Alternatively, if you think that having yet another recursive pass would be slow, we could use it from within `HighLevelSynthesis` itself, i.e. when `HighLevelSynthesis` needs to synthesize a `DAGCircuit` (possibly as part of its own recursion), it can first call `OptimizeAnnotated` on this dag circuit with ``recurse=False``. What do you think?

Reply:
Yeah, this logic makes sense to me. I think leaving this as a standalone pass makes sense, just as it seems like a slightly different building block in a pipeline that high level synthesis. However, I do agree there is probably some concern around performance if we're going to run with `recurse=True` in the `init` stage and I like your suggestion of calling it internally when it makes sense. I think when we investigate integrating this into the preset pass managers as part of 1.1.0 we can look at all those details.